src/rules/procedure.rs

```diff
@@ -134,58 +135,49 @@ impl RuleDefinition for ReplaceEpilogue {
         node: &SyntaxNode,
         _ctx: &DboAnalyzeContext,
     ) -> Result<Vec<TextRange>, RuleError> {
-        if let Some(procedure) = Procedure::cast(node.clone()) {
-            let mut locations = find_children_tokens(procedure.syntax(), |t| {
-                // find an `END` keyword
-                if !(t.kind() == SyntaxKind::Keyword
-                    && t.text().to_string().to_lowercase() == "end")
-                {
-                    return false;
-                }
+        let locations = find_children_tokens(node, |t| {
+            // find an `END` keyword
+            if !(t.kind() == SyntaxKind::Keyword && t.text().to_string().to_lowercase() == "end") {
+                return false;
+            }
 
-                // determine if the `LANGUAGE` epilogue has already been applied
-                t.next_sibling_or_token()
-                    .and_then(|t| match t.kind() {
-                        SyntaxKind::SemiColon => t.next_sibling_or_token(),
-                        _ => None,
-                    })
-                    .and_then(|t| match t.kind() {
-                        SyntaxKind::Whitespace => t.next_sibling_or_token(),
-                        _ => None,
-                    })
-                    .and_then(|t| match t.kind() {
-                        SyntaxKind::DollarQuote => t.next_sibling_or_token(),
-                        _ => None,
-                    })
-                    .and_then(|t| match t.kind() {
-                        SyntaxKind::Whitespace => t.next_sibling_or_token(),
-                        _ => None,
-                    })
-                    .and_then(|t| match t.kind() {
-                        SyntaxKind::Ident => Some(t.to_string().to_lowercase() == "language"),
-                        _ => None,
-                    })
-                    .is_none()
-            })
-            .map(|t| {
-                let start = t.clone();
-                let end = t
-                    .siblings_with_tokens(Direction::Next)
-                    .find(|t| t.kind() == SyntaxKind::IdentGroup)
-                    .unwrap_or(Token(t));
-                TextRange::new(start.text_range().end(), end.text_range().end())
-            })
-            .collect::<Vec<TextRange>>();
+            // determine if the `LANGUAGE` epilogue has already been applied
+            t.parent()
+                .unwrap()
```

Comment:
I think it's better to refactor this function in a separate Pull Request to eliminate the `unwrap` call. Generally it should be avoided in code, there are only a few places where it makes sense, e.g. in tests or in CLI apps. If the `unwrap` fails the program will panic & exit.

One possible way to solve this is to ignore this particular case when a SyntaxToken `t` has no parent using `match` statement.

```rust
match t.parent() {
    Some(parent) => {
        parent.next_sibling_or_token()
            .and_then(...)
           // ...
           .is_none()
    }
    None => false,
}
```

In case the `unwrap` call is an error & should be propagated it's better to use `filter_map` in the appropriate place and collect using `collect::<Result<Vec<_>, _>>()`, that causes the iterator to stop at the first `Err`. It requires mapping items in an iterator chain to a `Result` type, but it allows the logic to keep `Err` around & use the `?`  operator. It's slightly more involved but I think that refactoring the iterator calls could also slightly improve readability.

Reply:
Good catch! I'll create a ticket for it.